### PR TITLE
Improve autocomplete for multiple search

### LINKED ISSUE #33: As a student, I can type part of the tag name and get the tag as search suggestions
... so that I can search faster and minimize the chance of an unsuccessful search.

For example:
`search t/fa` will show tags starting with `fa`.

### LINKED ISSUE #119: Feature: autocomplete for multiple search

### LINKED ISSUE #120: Remove find command
`find` command is now replaced with `search` command.

### LINKED ISSUE #121: Bug: Autocomplete box does not change with the SoConnect's window size

### PR DESCRIPTION
**Pull request by Ugholaf**
# Changes
- Autocomplete supports multiple search prefix search
- Removed find related commands
- Fix the error where the autocomplete box does not change when the application window changes

# Potential Bugs
1. Currently, when you press the down arrow key to select the autocomplete string, it does not update the input in the command box. Not sure if this will be counted as a bug, I have been trying to search online for the past few days, but can't seem to find any answer, maybe will leave this to fix in v1.4.
2. When the autocomplete string is selected (i.e highlighted), pressing the space key will trigger the application to do a search based on that autocomplete string instead of adding a space to the user input.

### CODE REVIEW

File: src/main/java/soconnect/commons/util/StringUtil.java [hunk 1]
```diff
@@ -102,6 +102,25 @@ public static boolean containsSomeKeywordsIgnoreCase(String sentence, String key
         return wordsInPreppedSentence.containsAll(preppedKeywords.subList(0, reducedKeywordsLength));
     }
 
+    /**
+     * Returns true if the {@code sentence} starts with the give {@code subString}, this ignores case.
```
Comment (florentianayuwono) [inline src/main/java/soconnect/commons/util/StringUtil.java]:
```suggestion
     * Returns true if the {@code sentence} starts with the given {@code subString}, this ignores case.
```

File: src/main/java/soconnect/commons/util/StringUtil.java [hunk 2]
```diff
@@ -102,6 +102,25 @@ public static boolean containsSomeKeywordsIgnoreCase(String sentence, String key
...
+     *   <br>examples:<pre>
+     *       containsWordIgnoreCase("ABc def", "ab") == true
```
Comment (florentianayuwono) [inline src/main/java/soconnect/commons/util/StringUtil.java]:
```suggestion
     *       startsWithIgnoreCase("ABc def", "ab") == true
```

File: src/main/java/soconnect/commons/util/StringUtil.java [hunk 3]
```diff
@@ -102,6 +102,25 @@ public static boolean containsSomeKeywordsIgnoreCase(String sentence, String key
...
+     *       containsWordIgnoreCase("ABc def", "Bc") == false
```
Comment (florentianayuwono) [inline src/main/java/soconnect/commons/util/StringUtil.java]:
```suggestion
     *       startsWithIgnoreCase("ABc def", "Bc") == false
```

File: src/main/java/soconnect/commons/util/StringUtil.java [hunk 4]
```diff
@@ -102,6 +102,25 @@ public static boolean containsSomeKeywordsIgnoreCase(String sentence, String key
...
+     *       containsWordIgnoreCase("ABc def", "de") == false
```
Comment (florentianayuwono) [inline src/main/java/soconnect/commons/util/StringUtil.java]:
```suggestion
     *       startsWithIgnoreCase("ABc def", "de") == false
```

File: src/main/java/soconnect/commons/util/StringUtil.java [hunk 5]
```diff
@@ -102,6 +102,25 @@ public static boolean containsSomeKeywordsIgnoreCase(String sentence, String key
...
+     *
+     * @param sentence Cannot be null.
+     * @param subString Cannot be null, cannot be empty.
+     */
+    public static boolean startWithIgnoreCase(String sentence, String subString) {
+        requireAllNonNull(sentence, subString);
+
+        String preppedSubString = subString.trim();
+        checkArgument(!subString.isEmpty());
+        return sentence.toLowerCase().startsWith(preppedSubString);
```
Comment (florentianayuwono) [inline src/main/java/soconnect/commons/util/StringUtil.java]:
```suggestion
        return sentence.toLowerCase().startsWith(preppedSubString.toLowerCase());
```

File: src/main/java/soconnect/logic/autocomplete/Autocomplete.java [hunk 6]
```diff
@@ -1,29 +1,71 @@
...
+     *   <br>examples:<pre>
+     *       If{@code argsString} = "and n/alex p/900", {@code prefix} = "n/" and "p/", this method returns
+     *       "p/900" as it is the last prefix with an argument.
+     *       If {@code argsString} = "and n/alex p/", {@code prefix} = "n/" and "p/", this method returns
+     *       and invalid prefix argument as the last prefix "p/" does not have an argument.
+     *       If {@code argsString} = "and n/alex p/", {@code prefix} = "e/" and "a/", this method returns
+     *       and invalid prefix argument as there is no prefix in the {@code argsString}.
+     *   </pre>
+     *
+     * @param argsString The arguments string of the form: {@code preamble <prefix>value <prefix>value ...}
```
Comment (florentianayuwono) [inline src/main/java/soconnect/logic/autocomplete/Autocomplete.java:49]:
```suggestion
     * @param argsString The arguments string of the form: {@code preamble <prefix>value <prefix>value ...}.
```

File: src/main/java/soconnect/logic/autocomplete/Autocomplete.java [hunk 7]
```diff
@@ -1,29 +1,71 @@
...
+     * @param prefixes Prefixes to tokenize the arguments string with.
+     * @return The string of the last prefix with argument if they are present, otherwise an invalid prefix argument.
+     */
+    String getLastPrefixArgument(String argsString, Prefix... prefixes);
+
+    /**
+     * Updates the filter of the filtered person list based on the preamble in the given {@code argsString}.
+     *
+     * @param argsString Arguments string of the form: {@code preamble <prefix>value <prefix>value...}
```
Comment (florentianayuwono) [inline src/main/java/soconnect/logic/autocomplete/Autocomplete.java:58]:
```suggestion
     * @param argsString Arguments string of the form: {@code preamble <prefix>value <prefix>value...}.
```

File: src/main/java/soconnect/logic/autocomplete/Autocomplete.java [hunk 8]
```diff
@@ -1,29 +1,71 @@
...
+     */
+    void updateFilteredPersonList(String argsString);
+
+    /**
+     * Generates a list of autocomplete entries.
+     *
+     * @param argsWithoutLastPrefixArgument The arguments string of the
+     *                                      form:{@code preamble <prefix>value <prefix>value ...}
```
Comment (florentianayuwono) [inline src/main/java/soconnect/logic/autocomplete/Autocomplete.java:66]:
```suggestion
     *                                      form:{@code preamble <prefix>value <prefix>value ...}.
```

File: src/main/java/soconnect/logic/autocomplete/AutocompleteManager.java [hunk 9]
```diff
@@ -1,47 +1,202 @@
...
+                    .map(person -> person.getTags().stream()
+                            .filter(tag -> startWithIgnoreCase(tag.tagName, argument))
+                            .map(tag -> autocompleteString + prefix + tag.tagName)
+                            .collect(Collectors.toList()))
+                    .flatMap(Collection::stream)
+                    .distinct()
+                    .limit(AUTOCOMPLETE_ENTRIES_LIMIT)
+                    .collect(Collectors.toList());
+        default:
+            return new ArrayList<>();
```
Comment (florentianayuwono) [inline src/main/java/soconnect/logic/autocomplete/AutocompleteManager.java:174]: The length is a bit concerning haha, but as long as you can defend yourself should be okay ;D

File: src/main/java/soconnect/ui/CommandBox.java [hunk 10]
```diff
@@ -72,57 +80,89 @@ private void handleCommandEntered() {
     }
 
     /**
-     * Displays a list of autocomplete entries.
-     * Solution below adapted from https://stackoverflow.com/questions/36861056/javafx-textfield-auto-suggestions.
+     * A Listener to execute autocomplete related actions.
```
Comment (florentianayuwono) [inline src/main/java/soconnect/ui/CommandBox.java]: I'm afraid the convention forces us to start with active verb :")

```suggestion
     * Activates a Listener to execute autocomplete related actions.
```

File: src/main/java/soconnect/ui/CommandBox.java [hunk 11]
```diff
@@ -72,57 +80,89 @@ private void handleCommandEntered() {
...
+            @Override
+            public void handle(MouseEvent event) {
+                autocompleteAction();
+            }
+        });
     }
 
     /**
-     * Generates a list of autocomplete entries with the given search result.
+     * The autocomplete action which will generate and display the autocomplete pop up.
```
Comment (florentianayuwono) [inline src/main/java/soconnect/ui/CommandBox.java]:
```suggestion
     * Generates and displays the autocomplete pop up.
```

File: src/main/java/soconnect/ui/CommandBox.java [hunk 12]
```diff
@@ -131,6 +171,47 @@ public void handle(ActionEvent event) {
...
+     */
+    private TextFlow buildTextFlow(String originalSearchInput, String autocompleteEntry) {
+        Text textBefore = new Text(autocompleteEntry.substring(0, originalSearchInput.length()));
+        Text textAfter = new Text(autocompleteEntry.substring(originalSearchInput.length()));
+        textBefore.setFill(Color.ORANGE);
+        return new TextFlow(textBefore, textAfter);
+    }
+
+    /**
+     * A Listener to detect the change in the application size and changes the label width accordingly.
```
Comment (florentianayuwono) [inline src/main/java/soconnect/ui/CommandBox.java]:
```suggestion
     * Activates a Listener to detect the change in the application size and changes the label width accordingly.
```

**Review by florentianayuwono — CHANGES_REQUESTED** [review]
Great work on the autocomplete!👍

File: src/main/java/soconnect/logic/autocomplete/Autocomplete.java (hunk 8 above)
Comment (Ugholaf) [reply src/main/java/soconnect/logic/autocomplete/Autocomplete.java:66]: From the AB3 documentation, they didn't include a full stop after the ellipses so I just follow. 

image

I tried adding the full stop and it looks like this which is quite ugly to me so I went to search about it.

Based on this website https://www.stylemanual.gov.au/grammar-punctuation-and-conventions/punctuation-and-capitalisation/ellipses#:~:text=Don't%20use%20a%20full,or%20semicolon%20after%20an%20ellipsis, we do not have to include a full stop after it.

File: src/main/java/soconnect/logic/autocomplete/AutocompleteManager.java (hunk 9 above)
Comment (Ugholaf) [reply src/main/java/soconnect/logic/autocomplete/AutocompleteManager.java:174]: Yup, I think it is too long, I will extract each of them out and make a private method for each indicator.

Comment (florentianayuwono) [reply src/main/java/soconnect/logic/autocomplete/AutocompleteManager.java:174]: The newest addition looks good!

**Review by florentianayuwono — APPROVED** [review]
LGTM!

**Pull request merged**